### ISSUE TITLE
Unexpected parameter count for sprintf with .= assignment

### ISSUE THREAD

**Issue by pelmered** [kalessil/phpinspectionsea#1188]
| Subject        | Details                                                                       |
| :------------- | :---------------------------------------------------------------------------- |
| Issue type     | Bug/False-Positive                          |
| Plugin         | Php Inspections (EA Extended) 3.0.10 |
| Language level | PHP 7.1                      |

Current behaviour
---
Screenshots of code + error message: 
https://prnt.sc/mdaa29
https://prnt.sc/mda9r3

**Code to reproduce:**
```php
        $var1 = 1;
        $var2 = 1;
        $var3 = 1;

        $messageFormat  = 'text';
        $messageFormat .= ' text %s text %s';
        $messageFormat .= ' text %s.';

        $message = sprintf(
            $messageFormat,
            $var1,
            $var2,
            $var3
        );
```

Expected behaviour
---

This should not cause an error or warning. If I assign the format string in one line, the error disappears. The original string is too long to fit on one line. 

Environment details
---
PHPStorm version 2018.3.3
MacOS 10.13.6

---

Thanks a lot for this awesome plugin!

**Comment by andreasschroth** [comment]
Why not doing this?

```php
$messageFormat  = 'text' .
        ' text %s text %s' .
        ' text %s.';
```

**Comment by pelmered** [comment]
@andreasschroth Sure that would work in this case, but it's still a bug isn't it?
If you for example want to make one of the parts conditional with an if statement there is no clean way to do that. 

For example:
```
$messageFormat  = 'text';
if(true) {
        $messageFormat .= ' text %s text %s';
} else {
        $messageFormat .= ' %s text %s';
}        
$messageFormat .= ' text %s.';

**Comment by rentalhost** [comment]
I guess that the initial case should be valid once that it is a simplified case of use, which proves that the bug exists. Maybe a workaround is just ignore this inspection in case of a special assignment (like `.=`) because will be very hard to inspection consolidate the final result of `$messageFormat` to identify that it have 3 terms (like this case).

**Comment by pelmered** [comment]
Yes, that would be ok. I can imagine that it would be very hard to figure out exactly what the contents of the variable is in this case. Especially with conditions like in my second example. Just ignoring would be fine for me.

**Comment by andreasschroth** [comment]
@pelmered Sure, obviously it's a bug that should be fixed or the inspection disabled in case the final result cannot be determined of the string. But I posted my suggestion in case you're not aware of the solution to at least avoid the inspection bug.

With your example though, I would still consider it a bad practise for myself. At least we work with localisation and concatenating strings that way would make any localisation impossible.

I would refactor your second example code to this:

```php
$messageFormat = 'text %s text %s text %s.';
if(true) {
        $messageFormat = 'text text %s text %s text %s.';
}
```

Those strings can then easily be passed to e.g. the gettext function `_()` for translation.

Nevertheless the inspection bug should be fixed though, I just don't see any good use to do it the way you described it in a real-world application.

**Comment by pelmered** [comment]
@andreasschroth Yes ok, I appreciate that. 
However, you make a lot of dangerous assumptions of what code in a real-world application can look like. There's as many ways to build software as there are developers, and when you make assumptions of how others should write their code you are on thin ice. 

I don't necessarily see a problem with translation here neither. You can either translate the individual parts separately, or you can pass the whole concatenated string to your translation method. The later option might cause some problems when you parse your code for translatable strings I would probably not recommend that and this discussion is not really relevant to this issue.

**Comment by andreasschroth** [comment]
@pelmered But your aware that the word order is not the same in all languages, right? And that there are languages from left to right and also from right to left? Etc.

I mean, I keep repeating myself that this bug should be fixed, but I still don't think this is a best practice to concatenate the such a string. And isn't this plug-in and the whole idea about writing good, high-quality code? This plugin and many inspections are biased and make a lot of assumptions.

Anyway, this is my last comment regarding your bug report. As I said, it should be fixed, but I gave you good advice now how to improve your code instead of concatenating your strings.

**Comment by andreasschroth** [comment]
@pelmered But you're aware that the word order is not the same in all languages, right? And that there are languages from left to right and also from right to left? Etc.

I mean, I keep repeating myself that this bug should be fixed, but I still don't think this is a best practice to concatenate such a string. And isn't this plug-in and the whole idea about writing good, high-quality code? This plugin and many inspections are biased and make a lot of assumptions. That's why there are settings and you can disable them individually, etc. So sure my feedback is a subjective one as well.

Anyway, this is my last comment regarding your bug report. As I said, obviously the bug in the inspection should be fixed, but I still don't see why anyone would prefer concatenating strings in that way. I meant my feedback in a friendly way - I wanted to be helpful and give feedback how to write it in a better way (also much easier for devs to read, just compare my example and your example with the if conditions).

**Comment by kalessil** [comment]
Obviously a bug, I'll take care of fixing it.

**Comment by kalessil** [comment]
Fixed!
